File: Algorithms/C++/00205-Isomorphic Strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        std::unordered_map<char, char> sMap;
        std::unordered_map<char, char> tMap;

        for(int i{0}; i < s.size(); i++){
            if(sMap.contains(s[i]) && sMap[s[i]] != t[i]){
                return false;
            } else if(tMap.contains(t[i]) && tMap[t[i]] != s[i]){
                return false;
            } else{
                sMap[s[i]] = t[i];
                tMap[t[i]] = s[i];
            }
        }

        return true;
    }
};

/*
    tc: O(n)
    sc: O(n)

    e -> a
    g -> d
    ------
    a -> e
    d -> g

    f -> b
    o -> a
    o -> r
    -----
    b -> f
    a -> o
    r -> o

    return false if we find that an existing key already maps to a different char.

*/

File: Algorithms/Python3/00205-Isomorphic Strings.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        sMap = {}
        tMap = {}
        size = len(s)
        for i in range(size):
            # check mapping from s to t
            if s[i] in sMap:
                if sMap.get(s[i]) is not t[i]:
                    return False
            else:
                sMap[s[i]] = t[i]
            # check mapping from t to s
            if t[i] in tMap:
                if tMap.get(t[i]) is not s[i]:
                    return False
            else:
                tMap[t[i]] = s[i]

        return True

'''
use two maps for a two-way mapping
sMap
e: a
g: d

tMap
a: e
g: d

sMap
f: b
o: a

'''